architecture-building-systems/CityEnergyAnalyst
# Several fixes to zone_helper to solve issues found after selecting fix_overlapping

This PR solves a few different issues I found when I tried to run the `zone_helper` again on an old case study while using the new `fix_overlapping` option. The test case was a site shapefile that generated a valid CEA case study months ago, but now caused the following issues:
1. `min_height` could not be converted to `int` because it was imported as a string, not a float (#3229)
2. after splitting the geometries of overlapping buildings, some geometry types that CEA cannot process were created (`LineString`, `MultiPolygon`, `GeometryCollection`) (#3231)

While fixing these issues I furthermore noted that the `zone_helper`, as currently implemented, will do the following with geometry types that cannot be processed by CEA:
1. If it's a LineString or Point, discard them (reasonable)
2. If it's a MultiPolygon, try to convert it to a Polygon by `unary_union`
3. If the result of the union is a Polygon, keep it
4. If the result of the union is a MultiPolygon, keep the first Polygon in the MultiPolygon and drop the rest
However, Step 4 can lead to arbitrarily dropping building geometries (#3231). The other Polygons may be useful parts of the building, and even if you did choose to keep only one, the current approach doesn't ensure that the most representative/useful part of the building is kept, just whatever is first in the list. This problem is exacerbated by the `fix_overlapping` option, as by removing parts of buildings it is highly likely that a MultiPolygon will be formed.

So this PR does the following:
1. convert `min_heights` and `min_levels` to floats before using them – this ensures that we are not passing unusable strings that will cause the script to fail (Resolves #3229)
2. pass the shapefile generated after fixing overlapping geometries back to the `flatten_geometries` function to ensure the resulting shapefile only includes `Polygons` (Resolves the first part of issue #3231)
3. change the `flatten_geometries` function to keep all Polygons in a MultiPolygon or GeometryCollection type while dropping Points and LineStrings (Resolves #3231 – the second part)

The resulting `flatten_geometries` method is much slower than the original one (0.432 s vs 5.921 s for my case study on my computer), but I believe this is a necessary improvement – and also, my case study is city-scale, so a few-second delay for such a large case study is unsurprising and in any case negligible compared to the computational expense for the `fix_overlapping` option.

To test:
1. create a case study with many different building geometries (you can try creating a case study the size of Singapore to replicate my exact case)
2. run the zone helper with `fix_overlapping = True`
3. verify that only Polygon shapes are exported

I'm requesting reviews from @MatNif since he wrote the `fix_overlapping` part of the script and @reyery since he wrote the `flatten_geometries` method.
